### PR TITLE
Added tests

### PR DESCRIPTION
**Pull request by CezaryKopacz**

### CODE REVIEW

**Comment by CezaryKopacz** [comment]
Can anyone from @RxSwiftCommunity/contributors check this?

**Comment by CezaryKopacz** [comment]
@RxSwiftCommunity/contributors Anyone :) ?

**Comment by ashfurrow** [comment]
Sorry about the delay here. I'm travelling at the moment but I'll taker a look asap unless @GuyKahlon beats me to it.

**Comment by freak4pc** [comment]
The only gotcha I have about this is that it's a bit weird having a unit test actually go out to the network. Even though it's echo.websocket.org, something still feels weird about it. Just my 2cents. :]

File: RxStarscreamTests/RxStarscreamTests.swift [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+extension WebSocketEvent: Equatable { }
+
+public func ==(lhs: WebSocketEvent, rhs: WebSocketEvent) -> Bool {
+    switch (lhs, rhs) {
+    case (.connected, .connected):
+        return true
+    case (.disconnected(_), .disconnected(_)):
+        return true
+    case (.message(_), .message(_)):
```
Comment (vhbit) [inline RxStarscreamTests/RxStarscreamTests.swift]: is there any reason to skip checking for message content equality?
the same below with `.data` and above with `.disconnected`

File: RxStarscreamTests/RxStarscreamTests.swift [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                }).addDisposableTo(disposeBag)
+
+        socket.rx.response
+                .subscribe(responseObserver)
+                .addDisposableTo(disposeBag)
+
+        waitForExpectations(timeout: 10.0) { error in
+
+            XCTAssertEqual(self.responseObserver.events.count, 2)
+            if case WebSocketEvent.message(let message) = self.responseObserver.events[1].value.element! {
```
Comment (vhbit) [inline RxStarscreamTests/RxStarscreamTests.swift]: `XCTAssertEqual(message, WebSocketEvent.message(sentMessage))` should be enough here with a proper `==` implementation, it'll make it a bit cleaner than having an `if case` imho

File: RxStarscreamTests/RxStarscreamTests.swift [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        let connectExpectation = expectation(description: "Connect")
+        let scheduler = TestScheduler(initialClock: 0)
+
+        connectedObserver = scheduler.createObserver(Bool.self)
+
+        let connected = socket.rx.connected.shareReplay(1)
+
+        connected.subscribe(onNext: { [unowned self] _ in
+                    self.socket.disconnect()
+                }).addDisposableTo(disposeBag)
```
Comment (vhbit) [inline RxStarscreamTests/RxStarscreamTests.swift]: As @freak4pc points in "sibling" PRs, that could be replaced with `dispose(by: disposeBag)` as `addDisposableTo` is going to be deprecated.

**Comment by CezaryKopacz** [comment]
Thanks all for the review, will fix this soon. @freak4pc - I have the same thoughts, so maybe I'll create a mocked echo service?

**Comment by CezaryKopacz** [comment]
@freak4pc @vhbit updated my PR - could you please check this again?

File: Cartfile.resolved [hunk 4]
```diff
@@ -1,2 +1,3 @@
-github "ReactiveX/RxSwift" "3.0.1"
-github "daltoniam/Starscream" "2.0.2"
+github "ReactiveX/RxSwift" "3.3.1"
```
Comment (freak4pc) [inline Cartfile.resolved]: Why did you choose specifically 3.3.1 here ?

File: RxStarscream.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -10,6 +10,19 @@
 		0B4DAAC71E0497BC00ECDBF9 /* RxCocoa.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 0B4DAAC61E0497BC00ECDBF9 /* RxCocoa.framework */; };
 		0B4DAAC81E04980000ECDBF9 /* RxCocoa.framework in Frameworks */ = {isa = PBXBuildFile; fileRef = 0B4DAAC61E0497BC00ECDBF9 /* RxCocoa.framework */; };
 		0B4DAAC91E04980000ECDBF9 /* RxCocoa.framework in Embed Frameworks */ = {isa = PBXBuildFile; fileRef = 0B4DAAC61E0497BC00ECDBF9 /* RxCocoa.framework */; settings = {ATTRIBUTES = (CodeSignOnCopy, RemoveHeadersOnCopy, ); }; };
+		BEF314431E86D13C00C15DD4 /* RxStarscreamTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = BEF314421E86D13C00C15DD4 /* RxStarscreamTests.swift */; };
```
Comment (freak4pc) [inline RxStarscream.xcodeproj/project.pbxproj:13]: Are all these new files copied to the test target? Don't have the time to review in the project itself but if you can verify there are no unneeded files in the RxStarscream project vs. the example project that would be great

File: RxStarscream.xcodeproj/project.pbxproj [hunk 6]
```diff
@@ -288,6 +415,48 @@
...
+				CLANG_WARN_SUSPICIOUS_MOVE = YES;
+				DEVELOPMENT_TEAM = 52N3Q8V7G5;
+				FRAMEWORK_SEARCH_PATHS = (
+					"$(inherited)",
+					"$(PROJECT_DIR)/Carthage/Build/iOS",
+				);
+				INFOPLIST_FILE = RxStarscreamTests/Info.plist;
+				IPHONEOS_DEPLOYMENT_TARGET = 10.2;
+				LD_RUNPATH_SEARCH_PATHS = "$(inherited) @executable_path/Frameworks @loader_path/Frameworks";
+				PRODUCT_BUNDLE_IDENTIFIER = com.cezarykopacz.RxStarscreamTests;
```
Comment (freak4pc) [inline RxStarscream.xcodeproj/project.pbxproj]: Probably better to change this to be compatible with the main project and not your own bundle identifier.

File: RxStarscream.xcodeproj/project.pbxproj [hunk 7]
```diff
@@ -288,6 +415,48 @@
...
```
Comment (freak4pc) [inline RxStarscream.xcodeproj/project.pbxproj]: Same here

File: RxStarscreamTests/RxStarscreamTests.swift [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    private var pongObserver: TestableObserver<WebSocketEvent>!
+    private var responseObserver: TestableObserver<WebSocketEvent>!
+    private var socket: WebSocket!
+
+    let disposeBag = DisposeBag()
+
+    override func setUp() {
+        super.setUp()
+
+        socket = WebSocket(url: URL(string: "wss://echo.websocket.org")!)
```
Comment (freak4pc) [inline RxStarscreamTests/RxStarscreamTests.swift:46]: So, nothing was changed here right? We're still hitting the network for this unit test? 
Did you specifically only change the `disposed(by:)` syntax here ? 
What other changes were done ? 

Thanks !

**Review by freak4pc — COMMENTED** [review]
Thanks for the hard work! 
Left some more minor questions . Also would love elaboration on what exactly was changed between this and the previous implementation as it seems it is still hitting the network etc. Was it just the `disposed(by:)` syntax?

File: Cartfile.resolved [hunk 9]
```diff
@@ -1,2 +1,3 @@
...
+github "RxSwiftCommunity/RxSwiftExt" "2.1.0"
```
Comment (freak4pc) [inline Cartfile.resolved]: Also can you explain why the dependency for RxSwiftExt here? Seems somewhat unneeded

File: RxStarscream.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -195,10 +293,15 @@
...
 			isa = PBXProject;
 			attributes = {
-				LastSwiftUpdateCheck = 0730;
+				LastSwiftUpdateCheck = 0820;
 				LastUpgradeCheck = 0730;
 				ORGANIZATIONNAME = "Guy Kahlon";
 				TargetAttributes = {
+					BEF3143F1E86D13C00C15DD4 = {
+						CreatedOnToolsVersion = 8.2.1;
+						DevelopmentTeam = 52N3Q8V7G5;
```
Comment (vhbit) [inline RxStarscream.xcodeproj/project.pbxproj]: It's better to have no team.

File: Cartfile.resolved (hunk 4 above)
Comment (CezaryKopacz) [reply Cartfile.resolved]: I didn't choose it, carthage updated it to the latest version compatible with ~3.0

File: RxStarscreamTests/RxStarscreamTests.swift (hunk 8 above)
Comment (CezaryKopacz) [reply RxStarscreamTests/RxStarscreamTests.swift:46]: No, we're not hitting the network. I simulated specific events through invoking delegate methods, for example:
`socket.delegate!.websocketDidConnect(socket: socket)`

I removed all code responsible for making network request & some minor fixes you suggested.

File: RxStarscream.xcodeproj/project.pbxproj (hunk 6 above)
Comment (CezaryKopacz) [reply RxStarscream.xcodeproj/project.pbxproj]: sure

File: RxStarscreamTests/RxStarscreamTests.swift (hunk 8 above)
Comment (freak4pc) [reply RxStarscreamTests/RxStarscreamTests.swift:46]: Oh gotya :) Cool. That's a nice idea. 
I'm thinking whether or not the force unwrapping is good here but I guess for a test case that should be fine.

File: RxStarscream.xcodeproj/project.pbxproj (hunk 10 above)
Comment (CezaryKopacz) [reply RxStarscream.xcodeproj/project.pbxproj]: 👍

File: Cartfile.resolved (hunk 9 above)
Comment (CezaryKopacz) [reply Cartfile.resolved]: I have no idea how it happened. In `Cartfile` there is no entry with `RxSwiftExt`. I'll fix this

**Comment by CezaryKopacz** [comment]
@freak4pc  PR updated

**Review by freak4pc — APPROVED** [review]
Looks good to me for merge 👍 
@vhbit Wanna look at this again before we merge?

**Comment by vhbit** [comment]
@freak4pc LGTM

**Comment by freak4pc** [comment]
@CezaryKopacz Thanks for all the hard work buddy! 🥇

**Comment by CezaryKopacz** [comment]
@freak4pc thanks for your time and tips!

**Comment by freak4pc** [comment]
Fixes #2

**Pull request merged**